### ISSUE TITLE
Setup Local AI Training and Fine-Tuning Infrastructure

### ISSUE THREAD

**Issue by Codelisk** [Codelisk/OwnAi#1]
# Feature: Local AI Training and Fine-Tuning Infrastructure Setup

**Type**: Feature Enhancement  
**Category**: AI/ML Infrastructure  
**Priority**: High  
**Complexity**: High

### Current System Environment
- **Platform**: Linux (Azure VM)
- **Architecture**: x86_64
- **CPU**: Intel Xeon Platinum 8370C @ 2.80GHz (2 cores)
- **RAM**: 7.8GB total, 6.5GB available
- **Storage**: 61GB available (124GB total, 52% used)
- **GPU**: None (CPU-only training)
- **Project State**: Empty directory structure (OwnAi folder ready for setup)

### Objective
Implement local AI training and fine-tuning capabilities on this Linux machine for programming-related AI tasks. The system must support:
1. Training custom AI models locally
2. Fine-tuning existing models
3. Integration with programming workflows
4. Resource-efficient operation within hardware constraints

---

### Hardware Limitations Assessment
- **No GPU**: CPU-only training significantly limits model size and training speed
- **8GB RAM**: Restricts model size to small-to-medium models (up to ~3B parameters with quantization)
- **2 CPU cores**: Training will be slow; inference-focused workflows recommended
- **61GB disk**: Sufficient for models, datasets, and framework installation

### Feasibility: ✅ POSSIBLE with constraints

CPU-only training is viable for:
- Small language models (< 3B parameters)
- Fine-tuning with LoRA/QLoRA techniques
- Specialized programming models
- Inference-heavy workflows with occasional fine-tuning

---

#### 1. Ollama (Local AI Runtime)
**Purpose**: Model serving, inference, and base model management  
**Why**: 
- Optimized for CPU inference
- Easy model management (llama3, codellama, phi, gemma, etc.)
- RESTful API for integration
- Low resource overhead
- Supports GGUF quantized models

**Installation**:
\`\`\`bash
curl -fsSL https://ollama.com/install.sh | sh
\`\`\`

#### 2. Unsloth (Fine-Tuning Framework)
**Purpose**: Memory-efficient fine-tuning  
**Why**:
- 2x faster training than standard methods
- 80% less memory usage (critical for 8GB RAM)
- Supports QLoRA for efficient fine-tuning
- Compatible with Hugging Face ecosystem
- Optimized for limited hardware

**Installation**:
\`\`\`bash
pip install "unsloth[colab-new] @ git+https://github.com/unslothai/unsloth.git"
pip install --no-deps xformers trl peft accelerate bitsandbytes
\`\`\`

#### 3. Supporting Tools
- **PyTorch** (2.0+): ML framework backend
- **Hugging Face Transformers**: Model architecture and tokenizers
- **PEFT** (Parameter-Efficient Fine-Tuning): LoRA/QLoRA adapters
- **BitsAndBytes**: 4-bit/8-bit quantization for memory efficiency
- **Datasets**: Data loading and preprocessing
- **TRL** (Transformer Reinforcement Learning): Training utilities

**Complete Python Environment**:
\`\`\`bash
pip install torch transformers datasets accelerate peft bitsandbytes trl sentencepiece
\`\`\`

---

### Option 2: LocalAI
- Self-hosted OpenAI alternative
- Good for inference, limited training support
- Recommendation: Consider for production serving only

### Option 3: llama.cpp + Python bindings
- Ultra-lightweight C++ inference
- No native training support
- Recommendation: Use for inference-only scenarios

### Option 4: GPT4All
- User-friendly desktop application
- No training/fine-tuning capabilities
- Recommendation: Not suitable for development needs

### Option 5: Custom PyTorch-Only Setup
- Maximum flexibility
- Requires more manual optimization
- Recommendation: Use Unsloth instead for efficiency

---

### Phase 1: Environment Setup (Priority 1)
**Location**: \`/home/daniel/OwnAi/OwnAi/\`

1. **Create project structure**:
\`\`\`bash
mkdir -p {models,datasets,training,scripts,configs}
\`\`\`

2. **Install Ollama system-wide**:
\`\`\`bash
curl -fsSL https://ollama.com/install.sh | sh
systemctl status ollama  # Verify service running
\`\`\`

3. **Setup Python virtual environment**:
\`\`\`bash
python3 -m venv venv
source venv/bin/activate
pip install --upgrade pip
\`\`\`

4. **Install ML dependencies**:
\`\`\`bash
pip install torch transformers datasets accelerate peft bitsandbytes trl sentencepiece
pip install "unsloth[colab-new] @ git+https://github.com/unslothai/unsloth.git"
pip install jupyter notebook  # For interactive development
\`\`\`

5. **Verify installations**:
\`\`\`bash
python -c "import torch; print(f'PyTorch: {torch.__version__}')"
python -c "import transformers; print(f'Transformers: {transformers.__version__}')"
python -c "from unsloth import FastLanguageModel; print('Unsloth OK')"
ollama --version
\`\`\`

### Phase 2: Model Setup (Priority 1)
**Location**: \`/home/daniel/OwnAi/OwnAi/models/\`

1. **Download base models via Ollama**:
\`\`\`bash
# Small coding model (1.3B parameters, ~2GB)
ollama pull deepseek-coder:1.3b

# Alternative: Phi-2 for general purpose (2.7B, ~3GB)
ollama pull phi

# Code-specific model
ollama pull codellama:7b-code
\`\`\`

2. **Download Hugging Face models for fine-tuning**:
\`\`\`python
# Script: scripts/download_base_model.py
from unsloth import FastLanguageModel

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="unsloth/tinyllama-1.1b-chat-bnb-4bit",
    max_seq_length=2048,
    dtype=None,  # Auto-detect
    load_in_4bit=True,  # Memory optimization
)
model.save_pretrained("./models/base-tinyllama-4bit")
tokenizer.save_pretrained("./models/base-tinyllama-4bit")
\`\`\`

### Phase 3: Fine-Tuning Pipeline (Priority 2)
**Location**: \`/home/daniel/OwnAi/OwnAi/training/\`

1. **Create training configuration**:
\`\`\`python
# configs/training_config.yaml
model:
  base_model: "unsloth/tinyllama-1.1b-chat-bnb-4bit"
  max_seq_length: 2048
  load_in_4bit: true

lora:
  r: 16
  lora_alpha: 16
  lora_dropout: 0
  target_modules: ["q_proj", "k_proj", "v_proj", "o_proj"]
  
training:
  per_device_train_batch_size: 2
  gradient_accumulation_steps: 4
  num_train_epochs: 3
  learning_rate: 2e-4
  fp16: true
  logging_steps: 10
  output_dir: "./training/outputs"
\`\`\`

2. **Create fine-tuning script template**:
\`\`\`python
# training/finetune.py
from unsloth import FastLanguageModel
from trl import SFTTrainer
from transformers import TrainingArguments
from datasets import load_dataset

# Load model with memory optimization
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="unsloth/tinyllama-1.1b-chat-bnb-4bit",
    max_seq_length=2048,
    dtype=None,
    load_in_4bit=True,
)

# Configure LoRA for efficient fine-tuning
model = FastLanguageModel.get_peft_model(
    model,
    r=16,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],
    lora_alpha=16,
    lora_dropout=0,
    bias="none",
    use_gradient_checkpointing="unsloth",
    random_state=3407,
)

# Load dataset (example: code dataset)
dataset = load_dataset("iamtarun/python_code_instructions_18k_alpaca", split="train")

# Training configuration
trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=dataset,
    dataset_text_field="text",
    max_seq_length=2048,
    args=TrainingArguments(
        per_device_train_batch_size=2,
        gradient_accumulation_steps=4,
        warmup_steps=5,
        num_train_epochs=3,
        learning_rate=2e-4,
        fp16=True,
        logging_steps=10,
        output_dir="./training/outputs",
        optim="adamw_8bit",
    ),
)

# Train
trainer.train()

# Save fine-tuned model
model.save_pretrained("./models/finetuned-coding-model")
tokenizer.save_pretrained("./models/finetuned-coding-model")
\`\`\`

3. **Create inference script**:
\`\`\`python
# scripts/inference.py
from unsloth import FastLanguageModel

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="./models/finetuned-coding-model",
    max_seq_length=2048,
    dtype=None,
    load_in_4bit=True,
)

FastLanguageModel.for_inference(model)

def generate_code(prompt):
    inputs = tokenizer(prompt, return_tensors="pt")
    outputs = model.generate(**inputs, max_new_tokens=512)
    return tokenizer.decode(outputs[0], skip_special_tokens=True)

# Test
result = generate_code("Write a Python function to calculate fibonacci numbers")
print(result)
\`\`\`

### Phase 4: Integration with Programming Workflow (Priority 3)
**Location**: \`/home/daniel/OwnAi/OwnAi/scripts/\`

1. **Create Ollama integration script**:
\`\`\`python
# scripts/ollama_helper.py
import requests
import json

class OllamaClient:
    def __init__(self, base_url="http://localhost:11434"):
        self.base_url = base_url
    
    def generate(self, model, prompt, stream=False):
        response = requests.post(
            f"{self.base_url}/api/generate",
            json={"model": model, "prompt": prompt, "stream": stream}
        )
        return response.json()
    
    def chat(self, model, messages):
        response = requests.post(
            f"{self.base_url}/api/chat",
            json={"model": model, "messages": messages}
        )
        return response.json()

# Usage
client = OllamaClient()
result = client.generate("deepseek-coder:1.3b", "Explain list comprehension in Python")
print(result['response'])
\`\`\`

2. **Create CLI tool for quick AI queries**:
\`\`\`python
# scripts/ai_cli.py
#!/usr/bin/env python3
import sys
from ollama_helper import OllamaClient

def main():
    if len(sys.argv) < 2:
        print("Usage: ai_cli.py 'your prompt here'")
        sys.exit(1)
    
    prompt = " ".join(sys.argv[1:])
    client = OllamaClient()
    result = client.generate("deepseek-coder:1.3b", prompt)
    print(result['response'])

if __name__ == "__main__":
    main()
\`\`\`

3. **Create model evaluation script**:
\`\`\`python
# scripts/evaluate_model.py
from transformers import pipeline
import time

def benchmark_inference(model_path, prompts):
    generator = pipeline("text-generation", model=model_path, device=-1)  # CPU
    
    results = []
    for prompt in prompts:
        start = time.time()
        output = generator(prompt, max_new_tokens=100)
        elapsed = time.time() - start
        results.append({"prompt": prompt, "time": elapsed, "output": output})
    
    return results
\`\`\`

### Phase 5: Dataset Preparation (Priority 2)
**Location**: \`/home/daniel/OwnAi/OwnAi/datasets/\`

1. **Create dataset preparation script**:
\`\`\`python
# scripts/prepare_dataset.py
from datasets import load_dataset, Dataset
import json

def prepare_coding_dataset():
    # Option 1: Use existing datasets
    datasets_to_try = [
        "iamtarun/python_code_instructions_18k_alpaca",
        "code_search_net",
        "codeparrot/github-code",
    ]
    
    # Option 2: Create custom dataset
    custom_data = []
    # Load from JSON, CSV, or text files
    # Format: {"instruction": "...", "input": "...", "output": "..."}
    
    return Dataset.from_list(custom_data)

def format_for_training(example):
    prompt = f"### Instruction:\n{example['instruction']}\n\n### Response:\n{example['output']}"
    return {"text": prompt}
\`\`\`

2. **Document dataset format**:
\`\`\`markdown
# datasets/README.md

## Dataset Format

For supervised fine-tuning, use this format:

{
  "instruction": "Write a function to reverse a string",
  "input": "",
  "output": "def reverse_string(s):\n    return s[::-1]"
}

For chat-based training:

{
  "messages": [
    {"role": "user", "content": "How do I reverse a string in Python?"},
    {"role": "assistant", "content": "You can use slicing: s[::-1]"}
  ]
}
\`\`\`

### Phase 6: Monitoring and Optimization (Priority 3)

1. **Create resource monitoring script**:
\`\`\`python
# scripts/monitor_training.py
import psutil
import time

def monitor_resources(interval=5):
    while True:
        cpu = psutil.cpu_percent(interval=1)
        mem = psutil.virtual_memory()
        print(f"CPU: {cpu}% | RAM: {mem.percent}% ({mem.used/1024**3:.1f}GB/{mem.total/1024**3:.1f}GB)")
        time.sleep(interval)
\`\`\`

2. **Create training logs parser**:
\`\`\`python
# scripts/parse_training_logs.py
import re
import matplotlib.pyplot as plt

def extract_loss_from_logs(log_file):
    losses = []
    with open(log_file) as f:
        for line in f:
            match = re.search(r"'loss': ([\d.]+)", line)
            if match:
                losses.append(float(match.group(1)))
    return losses

def plot_training_progress(losses):
    plt.plot(losses)
    plt.xlabel("Training Step")
    plt.ylabel("Loss")
    plt.title("Training Progress")
    plt.savefig("training_progress.png")
\`\`\`

---

## File Structure After Implementation

\`\`\`
/home/daniel/OwnAi/OwnAi/
├── venv/                          # Python virtual environment
├── models/                        # Stored models
│   ├── base-tinyllama-4bit/
│   ├── finetuned-coding-model/
│   └── ollama-models/            # Symlink to ~/.ollama/models
├── datasets/                      # Training datasets
│   ├── coding-instructions/
│   ├── custom-data/
│   └── README.md
├── training/                      # Training scripts and outputs
│   ├── finetune.py
│   ├── outputs/
│   └── checkpoints/
├── scripts/                       # Utility scripts
│   ├── download_base_model.py
│   ├── inference.py
│   ├── ollama_helper.py
│   ├── ai_cli.py
│   ├── evaluate_model.py
│   ├── prepare_dataset.py
│   ├── monitor_training.py
│   └── parse_training_logs.py
├── configs/                       # Configuration files
│   ├── training_config.yaml
│   └── model_config.json
├── notebooks/                     # Jupyter notebooks for experimentation
│   ├── exploration.ipynb
│   └── fine_tuning_experiments.ipynb
├── tests/                         # Test scripts
│   └── test_inference.py
├── requirements.txt               # Python dependencies
├── README.md                      # Project documentation
└── .gitignore
\`\`\`

---

### Training Speed (CPU-only, 8GB RAM)
- **TinyLlama 1.1B**: ~2-5 minutes per epoch (1000 samples)
- **Phi-2 2.7B**: ~10-20 minutes per epoch (1000 samples)
- **Larger models**: May require further quantization or external training

### Inference Speed
- **TinyLlama 1.1B**: ~5-10 tokens/second (CPU)
- **Phi-2 2.7B**: ~2-5 tokens/second (CPU)
- **Ollama GGUF models**: ~10-20 tokens/second (optimized)

### Memory Usage
- **Training**: 4-6GB RAM (with 4-bit quantization + LoRA)
- **Inference**: 2-4GB RAM (quantized models)
- **Buffer**: 2-4GB for system operations

---

## Risks and Mitigation

| Risk | Impact | Mitigation |
|------|--------|------------|
| OOM during training | High | Use 4-bit quantization, reduce batch size to 1, use gradient checkpointing |
| Slow training speed | Medium | Use small models (<3B), limit dataset size, consider cloud training for large jobs |
| Poor model quality | Medium | Start with pre-trained coding models, use high-quality datasets, iterate on prompts |
| Disk space exhaustion | Low | Monitor usage, use model compression, clean old checkpoints |
| CPU bottleneck | Medium | Use Ollama for inference (optimized), batch requests, consider model distillation |

---

## Validation Criteria

✅ **Success Metrics**:
1. Ollama service running and accessible at localhost:11434
2. Python environment with all dependencies installed
3. Successfully load and run inference on base model
4. Complete one full fine-tuning cycle without OOM errors
5. Fine-tuned model shows measurable improvement on test prompts
6. Inference latency < 10 seconds for 100-token generation
7. Training completes within 1 hour for 1000-sample dataset

---

## Cost Analysis

**Total Storage**: ~15-20GB
- Ollama models: 5-10GB
- Python environment: 2-3GB
- Training checkpoints: 2-5GB
- Datasets: 1-2GB

**Time Investment**:
- Initial setup: 2-3 hours
- First training run: 1-2 hours
- Learning curve: 5-10 hours

**Resource Cost**: $0 (all open-source)

---

## Next Steps for AI Agent

1. **Execute Phase 1**: Create project structure and install dependencies
2. **Execute Phase 2**: Download and verify base models
3. **Execute Phase 3**: Create training scripts from templates
4. **Execute Phase 4**: Setup Ollama integration and CLI tools
5. **Test inference**: Verify all models load and generate outputs
6. **Prepare sample dataset**: Create or download programming dataset
7. **Run test training**: Execute one epoch to validate pipeline
8. **Document results**: Record performance metrics and issues
9. **Iterate**: Optimize based on initial results

---

## References and Resources

**Documentation**:
- Ollama: https://ollama.com/
- Unsloth: https://github.com/unslothai/unsloth
- Hugging Face PEFT: https://huggingface.co/docs/peft
- TRL Library: https://huggingface.co/docs/trl

**Recommended Models**:
- unsloth/tinyllama-1.1b-chat-bnb-4bit (Best for CPU training)
- microsoft/phi-2 (2.7B, good balance)
- deepseek-ai/deepseek-coder-1.3b-base (Code-specific)

**Datasets**:
- iamtarun/python_code_instructions_18k_alpaca
- code_search_net
- bigcode/starcoderdata

---

## Conclusion

✅ **FEASIBLE**: Local AI training and fine-tuning is possible on this system with constraints.

**Recommended Approach**: 
1. Use **Ollama** for production inference (fast, optimized)
2. Use **Unsloth + PyTorch** for fine-tuning (memory-efficient)
3. Focus on models ≤ 3B parameters
4. Use 4-bit quantization and LoRA for training
5. Leverage pre-trained coding models

**Limitations**: 
- No GPU limits model size and training speed
- Best suited for small models and incremental fine-tuning
- Production workloads may benefit from cloud GPU training

**Viability Score**: 7/10 for development, 5/10 for production training

The system is ready for implementation with the above architecture.
